_posts/2021-02-04-behind-the-scenes-devise.markdown

```diff
@@ -0,0 +1,162 @@
+---
+layout: post
+title: "Behind The Scenes: Devise"
+date: 2021-02-04 14:00:00
+categories: ["learning", "devise"]
+author: arieljuod
+---
+
+**Devise** is a well known solution for authentication in **Rails** applications. It's full featured (not only adds authentication but also password recovery, email changing, session timeout, locking, ip tracking, and more) and can be expanded to add even more (like JWT authentication).
+
+In this post, I'll go over the code related to the basic authentication process, how it relates to **Warden** and some of the magic behind it.
+
+<!--more-->
+
+## Gem Initialization Process
+
+Like most gems, when loaded, the entry point is `lib/devise.rb`. Inside this file we can find a lot of code:
+- it requires many modules (like `rails`, `activesupport`, etc)
+- then it defines the Devise module which does many things:
+  - autoloads many modules for Controllers, Mailers, Tests, Strategies (we'll get back to this later)
+  - defines all the module's attribute accessors for configuration
+  - defines methods for extending devise
+  - and more
+- at the end it requires `warden` and Devise's extensions for `models`, `rails`, etc
+
+> https://github.com/heartcombo/devise/blob/master/lib/devise.rb
+
+If we take a look into the files in that last part of the `lib/devise.rb` file, we can understand better how it works.
+
+At `lib/devise/rails.rb`, we can see Devise runs as a Rails Engine and it adds Warden as a middleware:
+
+```ruby
+# Initialize Warden and copy its configurations.
+config.app_middleware.use Warden::Manager do |config|
+  Devise.warden_config = config
+end
+```
+
+At `lib/devise/models.rb`, we can see it defines the Device::Models module that which is later included in ActiveRecord::Base and adds the `devise` helper method used to configure our authenticable ActiveRecord models:
+
+```ruby
+module Devise
+  module Models
+    def devise(*modules)
+      options = modules.extract_options!.dup
+
+      selected_modules = modules.map(&:to_sym).uniq.sort_by do |s|
+        Devise::ALL.index(s) || -1  # follow Devise::ALL order
+      end
+
+      ...
+```
+
+```ruby
+# lib/devise/orm/active_record.rb
+
+require 'orm_adapter/adapters/active_record'
+
+ActiveSupport.on_load(:active_record) do
+  extend Devise::Models
+end
+```
+
+And we can use it like:
+
+```ruby
+class User < ApplicationRecord
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+end
+```
+
+## What's Warden?
+
+Warden is a gem, added as a middleware by Devise, that takes care of fetching the authentication information from the request and fetching the user. How it does that depends on the Strategy we use for authentication. The default strategy is `database_authenticatable` [link](https://github.com/heartcombo/devise/blob/master/lib/devise/strategies/database_authenticatable.rb), but it can be changed using the model's `devise` helper method we saw above.
```

Comment:
```suggestion
Warden is a gem that takes care of fetching authentication details from the request and fetching the user. Devise requires Warden to be added as a middleware. How it does that depends on the Strategy we use for authentication. The default strategy is `database_authenticatable` [link](https://github.com/heartcombo/devise/blob/master/lib/devise/strategies/database_authenticatable.rb), but it can be changed using the model's `devise` helper method we saw above.
```